homeassistant/components/input_datetime.py

```diff
@@ -0,0 +1,214 @@
+                input_datetime.async_set_datetime(
+                    call.data.get(ATTR_DATE, None),
+                    call.data.get(ATTR_TIME, None)
+                )
+            )
+
+        if tasks:
+            yield from asyncio.wait(tasks, loop=hass.loop)
+
+    hass.services.async_register(
+        DOMAIN, SERVICE_SET_DATETIME, async_set_datetime_service,
+        schema=SERVICE_SET_DATETIME_SCHEMA)
+
+    yield from component.async_add_entities(entities)
+    return True
+
+
+class DatetimeSelect(Entity):
+    """Representation of a select datetime."""
+
+    def __init__(self, object_id, name, has_date, has_time, icon):
+        """Initialize a select input."""
+        self.entity_id = ENTITY_ID_FORMAT.format(object_id)
+        self._name = name
+        self._has_date = has_date
+        self._has_time = has_time
+        self._icon = icon
+        self._current_datetime = None
+
+    @asyncio.coroutine
+    def async_added_to_hass(self):
+        """Run when entity about to be added."""
+        if self._current_datetime is not None:
+            return
+
+        old_state = yield from async_get_last_state(self.hass, self.entity_id)
+        if old_state is not None:
+            restore_val = dt_util.parse_datetime(old_state.state)
+        else:
+            restore_val = dt_util.now()
+
+        if not self._has_date:
+            self._current_datetime = restore_val.time()
+        elif not self._has_time:
+            self._current_datetime = restore_val.date()
+        else:
+            self._current_datetime = restore_val
+
+    @property
+    def should_poll(self):
+        """If entity should be polled."""
+        return False
+
+    @property
+    def name(self):
+        """Return the name of the select input."""
+        return self._name
+
+    @property
+    def icon(self):
+        """Return the icon to be used for this entity."""
+        return self._icon
+
+    @property
+    def state(self):
+        """Return the state of the component."""
+        if self._current_datetime is None:
+            return None
+
+        print("Returning type: " + str(type(self._current_datetime)))
+        return self._current_datetime
+
+    @property
+    def state_attributes(self):
+        """Return the state attributes."""
+        attrs = {
+            'has_date': self._has_date,
+            'has_time': self._has_time,
+        }
+
+        if self._has_date and self._current_datetime is not None:
+            attrs['year'] = self._current_datetime.year
+            attrs['month'] = self._current_datetime.month
+            attrs['day'] = self._current_datetime.day
+        if self._has_time and self._current_datetime is not None:
+            attrs['hour'] = self._current_datetime.hour
+            attrs['minute'] = self._current_datetime.minute
+            attrs['second'] = self._current_datetime.second
+
+        return attrs
+
+    @callback
+    def async_set_datetime(self, date_val, time_val):
```

Comment:
Don't validate stuff inside this function. That will be done with vololuptus

Reply:
The problem is that inside the service schema, I need to make date and time both optional, because there can be `input_datetime`s that dont carry date / time. But, depending on the actual configuration of the entity, they need time and / or date. I don't know that at the time I register the service and therefore the service schema.

Thus, I cannot assume that either `date_val` or `time_val` is not `None` in this method.

Another option I see would be to break the service into three: SET_DATETIME, SET_DATE and SET_TIME, which then have required 'date' and/or 'time' data associated with them. Is that a better solution? It would mean that callers such as the frontend need a bit more logic to figure out which service to call…

Reply:
Wouldn't you be able to use the `has_at_least_one_key` validation then? https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/helpers/config_validation.py#L45

Reply:
@balloob You mean in the schema for `SET_DATETIME` without breaking it into three? The problem is that the question which of `CONF_HAS_TIME` and/or `CONF_HAS_DATE` are set to `True` determines exactly which keys must be there. For example: Say you configure the input like this:

```
… 
  has_time: True
  has_date: False
…
```

and you call the service with `{'date': '2017-01-01'}`. Now the service schema with `has_at_least_one_key(['date', 'time'])` happily validates, but it's the service can't work with this data.

Reply:
In that case you can use `vol.Any` ([example](https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/helpers/config_validation.py#L190-L191))